state/machine.go

```diff
@@ -412,6 +385,84 @@ func (machine *machine) buildSignedBlock() block.SignedBlock {
 	return signedBlock
 }
 
+func (machine *machine) preconditionCheck() {
+	if machine.lockedRound < 0 {
+		if machine.lockedValue != nil {
+			panic("expected locked block to be nil")
+		}
+	}
+	if machine.lockedRound >= 0 {
+		if machine.lockedValue == nil {
+			panic("expected locked block to not be nil")
+		}
+	}
+
+	if machine.validRound < 0 {
+		if machine.validValue != nil {
+			panic("expected valid block to be nil")
+		}
+	}
+	if machine.validRound >= 0 {
+		if machine.validValue == nil {
+			panic("expected valid block to not be nil")
+		}
+	}
+}
+
+func (machine *machine) handleTransitionsForHigherRounds(transition Transition) Action {
+	if transition.Round() > machine.currentRound {
+		if _, ok := machine.bufferedMessages[transition.Round()]; !ok {
+			machine.bufferedMessages[transition.Round()] = map[sig.Signatory]struct{}{}
+		}
+		machine.bufferedMessages[transition.Round()][transition.Signer()] = struct{}{}
+	}
+
+	higherRound := machine.checkForHigherRounds()
+	if higherRound > machine.currentRound {
+		// Found f+1 messages for a higher round, progress to the new round.
+		switch transition := transition.(type) {
+		case PreVoted:
+			// At this stage, we want to buffer all prevotes. It doesn't matter if
+			// the prevote is new or not, because we want to start a new round regardless.
+			_ = machine.polkaBuilder.Insert(transition.SignedPreVote)
+		case PreCommitted:
+			// At this stage, we want to buffer all precommits. It doesn't matter if
+			// the precommit is new or not, because we want to start a new round regardless.
+			_ = machine.commitBuilder.Insert(transition.SignedPreCommit)
+		}
+		return machine.StartRound(higherRound, nil)
+	}
+	return nil
+}
+
+func (machine *machine) handleTimers(tick Ticked) Action {
```

Comment:
I think this does most of the work towards isolating the timeout logic as much as possible, but I think a few more changes would be good. I think `handleTimers` should only be responsible for timeout logic, and should basically look like
```golang
if machine.proposeTimer.AcceptTick(tick) {
	return machine.onTimeoutPropose(machine.currentRound)
}
if machine.prevoteTimer.AcceptTick(tick) {
	return machine.onTimeoutPrevote(machine.currentRound)
}
if machine.precommitTimer.AcceptTick(tick) {
	return machine.onTimeoutPrecommit(machine.currentHeight, machine.currentRound)
}
return nil
```
with the logic put inside the new methods `onTimeout*`. I think this is easier to understand and also more closely reflects the paper, which will make it easier to verify correctness.

As for the timer activation logic, I think this should go inside `Transition`. This is currently true for the propose timer, but for the other timers I think it would be best that we simply `switch` on the transition type and when it is a prevote we call (a suitably modified) `checkAndSchedulePrevoteTimeout` and when it is a precommit we call (a suitably modified) `checkAndSchedulePrecommitTimeout`, since these transition types are what can activate the timers, with one exception: for the precommit timer we would also need to run the check whenever the state transitions to the prevote state.

Reply:
Based on our discussion, I have updated this.